packages/sanity/src/core/studio/DevServerStopped.tsx

```diff
@@ -0,0 +1,74 @@
+/* eslint-disable i18next/no-literal-string -- will not support i18n in error boundaries */
+import {Card, Container, Heading, Stack, Text} from '@sanity/ui'
+import {type ReactNode, useEffect, useRef, useState} from 'react'
+
+const ERROR_TITLE = 'Dev server stopped'
+const ERROR_DESCRIPTION =
+  'The development server has stopped. You may need to restart it to continue working.'
+
+class DevServerStoppedError extends Error {
+  isDevServerStoppedError: boolean
+
+  constructor() {
+    super(ERROR_TITLE)
+    this.name = 'DevServerStoppedError'
+    this.isDevServerStoppedError = true
+  }
+}
+
+const useDetectDevServerStopped = () => {
+  const [devServerStopped, setDevServerStopped] = useState(false)
+  const serverIsReadyRef = useRef(false)
+
+  useEffect(() => {
+    const url = `ws://${window.location.hostname}:${window.location.port}/`
```

Comment:
Could it be better to [reuse Vite's existing WebSocket connection](https://vitejs.dev/guide/api-hmr#hot-on-event-cb) by utilising its `vite:ws:disconnect` event? It can be subscribed to like this: `import.meta.hot.on('vite:ws:disconnect', () => {})`.

Espen uses events like this in his [use-hot-module-reload](https://github.com/rexxars/use-hot-module-reload/blob/main/use-hot-module-reload/src/use-hot-module-reload.mts#L81) package, which Studio already consumes.

This also got me thinking: `useHotModuleReload` is already used inside `packages/sanity/src/core/studio/StudioErrorBoundary.tsx`. Could there be an argument to modify the `use-hot-module-reload` package so that it returns a state object containing the connection state?

Reply:
I did try using `import.meta.hot.on` but could never get any events to trigger. 

But I can look into it again now I know it's being used elsewhere

Reply:
@juice49 looked back into this, I think because I'd previously experimented with `import.meta.hot.on` inside lazily loaded modules, that was why I wasn't finding it to work. Can confirm it does work nicely.

I've updated this hook to use that approach.

Agreed, this would be nice in the `use-hot-module-reload` - I'll create a separate task for this since it looks like there are some tech debt improvements that can be made in that package for later versions of vite (and I'll also need to follow up with Espen to get this included and then released). Don't want to block this change right now. Hope that sounds ok with you?

Reply:
Awesome!! 🥳 Sounds great to me.